### DISCUSSION TITLE
NI X-series digital edge counting

### DISCUSSION THREAD

**Discussion by gadamc** [Ulm-IQO/qudi-core#24] [Q&A] [answered]
I have two questions regarding the ni x-series finite sampling / stream code. 

1. When configuring the clock task, why configure the clock to [output to a counter](https://github.com/Ulm-IQO/qudi-iqo-modules/blob/c650a61300cfeec51c4fbf8b91bdf3c9ab0ddcf7/src/qudi/hardware/ni_x_series/ni_x_series_finite_sampling_input.py#L499)?  Just to have that clock signal externally available? 
2. Can you explain how the [digital counting is configured](https://github.com/Ulm-IQO/qudi-iqo-modules/blob/c650a61300cfeec51c4fbf8b91bdf3c9ab0ddcf7/src/qudi/hardware/ni_x_series/ni_x_series_finite_sampling_input.py#L585-L617)?  This might be less of a Qudi question and more of an NI question. I'm very new to using NI DAQmx. But, my read of the Qudi code is that a digital input task is created and then the code
   
   * add_ci_period_chan -- add a counter to measure the period of a digital signal
   * DAQmxSetCIPeriodTerm -- set the terminal to measure the period to be the terminal of the clock!
   * DAQmxSetCICtrTimebaseSrc -- set the timebase of the counter to be controlled by the digital input channel. 

I'm having trouble understanding what is actually being measured here. It reads like qudi is measuring the period of the clock, but only when the digital input from the photon detector is high.  However, the results I'm getting seem to be physically correct. That is, we clearly see the 2.87 GHz PL drop when doing a CW ODMR using qudi. So, it's measuring something that correlates to PL intensity.

I've followed an [approach outlined here](https://nspyre.readthedocs.io/en/latest/guides/ni-daqmx.html) that seems intuitively correct to to me, but I'm measuring ~3x higher rate and I don't understand why. So you don't have to read that page, the steps taken in that approach are

  * clock_task.di_channels.add_di_chan('Dev1/port0')
  * detector_task.ci_channels.add_ci_count_edges_chan('Dev1/ctr2')
  * detector_task.ci_channels[0].ci_count_edges_term = '/Dev1/PFI12'
  * detector_task.timing.cfg_samp_clk_timing(source = 'Dev1/di/SampleClock')

Can you explain the difference and why Qudi's implementation actually works? 

Thank you!

**Comment by gadamc** [comment, marked as answer]
I've answered my own question (#2) above. Configuring the edge counter with the period measurement function, setting the output of the photon counter as the time source / clock and setting the clock to be the input signal effectively turns the edge counter into a rate counter. 

Thanks again.

**Reply by Neverhorst** [reply]
Thank you for reporting your findings. It's very helful for people who might have the same question.
